File: packages/sdk/src/runtime.ts
import { SanctumClient, type SanctumClientOptions } from './client.js'
import {
  AgentMemoryVault,
  resolveMemoryKey,
  type AgentMemoryVaultOptions,
} from './agent-memory.js'
import { MarketplaceClient } from './marketplace.js'
import {
  ControlPlaneSession,
  type ConnectOptions,
  type ConnectResult,
  type RegisterAgentOptions,
} from './control-plane.js'
import { attachSanctumRuntime, createSanctumMiddleware } from './middleware.js'
import type { ActionPolicy, ActionRequest, ActionResult, PolicyMap } from './types.js'
import type { VerificationStatus } from './verification.js'

export type PolicyMode = 'approve' | 'verify' | 'block'

function policyModeToPatch(mode: PolicyMode): Partial<ActionPolicy> {
  switch (mode) {
    case 'approve':
      return { requiresVerification: false, autoBlock: false }
    case 'verify':
      return { requiresVerification: true, autoBlock: false }
    case 'block':
      return { requiresVerification: false, autoBlock: true }
  }
}

/** Runtime entry — embed in agents, backends, robotics hosts. */
export class SanctumRuntime {
  private client: SanctumClient
  readonly control: ControlPlaneSession
  readonly marketplace: MarketplaceClient

  constructor(options: SanctumClientOptions = {}) {
    this.client = new SanctumClient(options)
    this.control = new ControlPlaneSession(this.client)
    this.marketplace = new MarketplaceClient(this.client)
  }

  /**
   * Install a marketplace package and connect using its template defaults.
   */
  async connectFromPackage(
    packageSlug: string,
    organizationId: string,
    overrides?: Partial<ConnectOptions>,
  ) {
    const { connect } = await this.marketplace.install(packageSlug, organizationId)
    const opts = this.marketplace.toConnectOptions(connect, organizationId)
    const result = await this.connect({ ...opts, ...overrides })
    for (const agent of connect.agents ?? []) {
      await this.registerAgent(agent)
    }
    return { ...result, packageSlug, packageVersion: connect.packageVersion }
  }

  /** Register this host with the Sanctum control plane (telemetry + heartbeat). */
  connect(options: ConnectOptions): Promise<ConnectResult> {
    return this.control.connect(options)
  }

  disconnect(): Promise<void> {
    return this.control.disconnect()
  }

  registerAgent(options: RegisterAgentOptions) {
    return this.control.registerAgent(options)
  }

  emitEvent(eventType: string, payload?: Record<string, unknown>, agentId?: string) {
    return this.control.emitEvent(eventType, payload, agentId)
  }

  /**
   * Encrypted memory vault for an agent. Ciphertext is stored in the control plane;
   * decryption keys never leave the host (SANCTUM_MEMORY_KEY).
   */
  /**
   * Re-wrap all encrypted memory for an agent under a new SANCTUM_MEMORY_KEY.
   */
  async rotateAgentMemoryKeys(
    agentId: string,
    newEncryptionKey: string,
    options?: { oldEncryptionKey?: string; salt?: string },
  ) {
    if (!this.control.runtimeId) throw new Error('Call connect() before rotateAgentMemoryKeys()')
    const oldKey = options?.oldEncryptionKey ?? resolveMemoryKey()
    const vault = this.memory(agentId, {
      encryptionKey: oldKey,
      salt: options?.salt,
    })
    const result = await vault.rotateKey(newEncryptionKey, {
      oldEncryptionKey: oldKey,
    })
    if (this.control.connected) {
      void this.emitEvent('memory.key_rotated', {
        agentId,
        rotated: result.rotated,
        failed: result.failed.length,
        newKeyHint: result.newKeyHint,
      })
    }
    return result
  }

  memory(
    agentId: string,
    options?: Pick<AgentMemoryVaultOptions, 'encryptionKey' | 'salt'>,
  ): AgentMemoryVault {
    if (!this.control.runtimeId) {
      throw new Error('Call connect() before memory()')
    }
    const org = this.control.organizationId
    return new AgentMemoryVault(this.client, this.control.runtimeId, agentId, {
      encryptionKey: resolveMemoryKey(options?.encryptionKey),
      salt: options?.salt ?? (org ? `sanctum:${org}:${agentId}` : undefined),
    })
  }

  async verifyAction(
    request: ActionRequest,
    options?: { offlineMode?: boolean; correlationId?: string },
  ): Promise<ActionResult> {
    const enriched: ActionRequest =
      this.control.organizationId != null
        ? {
            ...request,
            context: { ...request.context, org_id: this.control.organizationId },
          }
        : request
    const result = await this.client.verifyAction(enriched, options)
    if (this.control.connected) {
      const type =
        result.decision === 'BLOCKED'
          ? 'policy.blocked'
          : result.decision === 'REQUIRE_VERIFICATION'
            ? 'policy.verification_required'
            : 'policy.approved'
      void this.control.emitEvent(type, {
        action: request.action,
        decision: result.decision,
        correlationId: result.correlationId,
        org_id: this.control.organizationId,
      }, request.actor)
    }
    return result
  }

  getPolicies(): Promise<PolicyMap> {
    return this.client.getPolicies()
  }

  getStatus() {
    return this.client.getStatus()
  }

  getAudit(limit?: number) {
    return this.client.getAudit(limit)
  }

  replayAudit(limit?: number, orgId?: string) {
    return this.client.replayAudit(limit, orgId)
  }

  getEvidenceSummary(limit?: number, orgId?: string) {
    return this.client.getEvidenceSummary(limit, orgId)
  }

  verifyActionToken(token: string) {
    return this.client.verifyActionToken(token)
  }

  getVerificationStatus(correlationId: string): Promise<VerificationStatus> {
    return this.client.getVerificationStatus(correlationId)
  }

  waitForVerification(
    correlationId: string,
    options?: { timeoutMs?: number; pollIntervalMs?: number },
  ): Promise<VerificationStatus> {
    return this.client.waitForVerification(correlationId, options)
  }

  resolveAuditEntry(
    id: string,
    body: {
      decision: 'APPROVED' | 'BLOCKED'
      resolvedBy?: string
      note?: string
    },
  ) {
    return this.client.resolveAuditEntry(id, body)
  }

  /** Set policy mode for an action — approve | verify | block. */
  policy(action: string, mode: PolicyMode): Promise<PolicyMap> {
    return this.client.updatePolicy(action, policyModeToPatch(mode))
  }

  /** Register any action name with optional policy fields (unlimited policies). */
  registerPolicy(
    action: string,
    modeOrPatch: PolicyMode | Partial<ActionPolicy> = 'approve',
  ): Promise<PolicyMap> {
    const patch =
      typeof modeOrPatch === 'string' ? policyModeToPatch(modeOrPatch) : modeOrPatch
    return this.client.createPolicy(action, patch)
  }

  deletePolicy(action: string): Promise<PolicyMap> {
    return this.client.deletePolicy(action)
  }

  exportPoliciesYaml(): Promise<string> {
    return this.client.exportPoliciesYaml()
  }

  importPoliciesYaml(yaml: string, merge = true): Promise<PolicyMap> {
    return this.client.importPoliciesYaml(yaml, merge)
  }

  getWebhookStatus() {
    return this.client.getWebhookStatus()
  }

  /** Agent middleware: `agent.use(sanctum.middleware())` pattern. */
  middleware() {
    return createSanctumMiddleware(this)
  }

  /** Robotics attach: `robot.attach(sanctum.runtime())` — Phase 1 registers onAction if present. */
  runtime() {
    return {
      attach: <T extends Parameters<typeof attachSanctumRuntime>[0]>(host: T) =>
        attachSanctumRuntime(host, this),
      middleware: () => this.middleware(),
    }
  }
}

File: packages/python-sdk/sanctum_runtime/runtime.py
from __future__ import annotations

from typing import Literal

from sanctum_runtime.client import SanctumClient
from sanctum_runtime.control_plane import ControlPlaneSession
from sanctum_runtime.errors import (
  SanctumActionBlockedError,
  SanctumApiError,
  SanctumVerificationRequiredError,
)
from sanctum_runtime.types import (
  ActionPolicy,
  ActionRequest,
  ActionResult,
  ActionTokenVerification,
  AuditReplayResult,
  EvidenceSummary,
  ExecutionStatus,
  PolicyMap,
  PolicyMode,
  SimulateResult,
)

PolicyMode = Literal["approve", "verify", "block"]


def _mode_to_policy(mode: PolicyMode) -> ActionPolicy:
  if mode == "approve":
    return {"requiresVerification": False, "autoBlock": False}
  if mode == "verify":
    return {"requiresVerification": True, "autoBlock": False}
  return {"requiresVerification": False, "autoBlock": True}


class SanctumRuntime:
  """High-level wrapper — mirrors SanctumRuntime in the Node SDK."""

  def __init__(self, client: SanctumClient | None = None, **client_kwargs: object) -> None:
    self._client = client or SanctumClient(**client_kwargs)  # type: ignore[arg-type]
    self.control = ControlPlaneSession(self._client)

  def connect(self, **kwargs: object) -> dict:
    return self.control.connect(**kwargs)  # type: ignore[arg-type]

  def disconnect(self) -> None:
    self.control.disconnect()

  def register_agent(self, **kwargs: object) -> object:
    return self.control.register_agent(**kwargs)  # type: ignore[arg-type]

  def emit_event(self, event_type: str, payload: dict | None = None, *, agent_id: str | None = None) -> object:
    return self.control.emit_event(event_type, payload, agent_id=agent_id)

  @property
  def client(self) -> SanctumClient:
    return self._client

  def verify_action(
    self,
    request: ActionRequest,
    *,
    raise_on_block: bool = True,
    raise_on_verify: bool = False,
    **kwargs: object,
  ) -> ActionResult:
    result = self._client.verify_action(request, **kwargs)  # type: ignore[arg-type]
    decision = result.get("decision")
    if self.control.connected:
      event_type = (
        "policy.blocked"
        if decision == "BLOCKED"
        else "policy.verification_required"
        if decision == "REQUIRE_VERIFICATION"
        else "policy.approved"
      )
      self.control.emit_event(
        event_type,
        {
          "action": request.get("action"),
          "decision": decision,
          "correlationId": result.get("correlationId"),
        },
        agent_id=request.get("actor"),
      )
    if raise_on_block and decision == "BLOCKED":
      raise SanctumActionBlockedError(result)
    if raise_on_verify and decision == "REQUIRE_VERIFICATION":
      raise SanctumVerificationRequiredError(result)
    return result

  def get_policies(self) -> PolicyMap:
    return self._client.get_policies()

  def get_status(self):
    return self._client.get_status()

  def get_audit(self, limit: int = 50, org_id: str | None = None) -> list[ActionResult]:
    return self._client.get_audit(limit, org_id=org_id)

  def simulate_action(
    self,
    request: ActionRequest,
    *,
    offline_mode: bool | None = None,
  ) -> SimulateResult:
    return self._client.simulate_action(request, offline_mode=offline_mode)

  def replay_audit(self, limit: int = 100, org_id: str | None = None) -> AuditReplayResult:
    return self._client.replay_audit(limit, org_id=org_id)

  def get_evidence_summary(self, limit: int = 200, org_id: str | None = None) -> EvidenceSummary:
    return self._client.get_evidence_summary(limit, org_id=org_id)

  def verify_action_token(self, token: str) -> ActionTokenVerification:
    return self._client.verify_action_token(token)

  def report_action_execution(
    self,
    entry_id: str,
    *,
    action_token: str,
    status: ExecutionStatus,
    reported_by: str | None = None,
    result_summary: str | None = None,
    output_ref: str | None = None,
    error: str | None = None,
    duration_ms: int | float | None = None,
  ) -> ActionResult:
    return self._client.report_action_execution(
      entry_id,
      action_token=action_token,
      status=status,
      reported_by=reported_by,
      result_summary=result_summary,
      output_ref=output_ref,
      error=error,
      duration_ms=duration_ms,
    )

  def policy(self, action: str, mode: PolicyMode) -> PolicyMap:
    return self._client.update_policy(action, _mode_to_policy(mode))

  def register_policy(self, action: str, mode: PolicyMode | ActionPolicy) -> PolicyMap:
    patch = _mode_to_policy(mode) if isinstance(mode, str) else mode
    try:
      return self._client.update_policy(action, patch)
    except SanctumApiError as e:
      if e.status_code == 404:
        return self._client.create_policy(action, patch)
      raise

  def delete_policy(self, action: str) -> PolicyMap:
    return self._client.delete_policy(action)

  def export_policies_yaml(self) -> str:
    return self._client.export_policies_yaml()

  def import_policies_yaml(self, yaml: str, merge: bool = True) -> PolicyMap:
    return self._client.import_policies_yaml(yaml, merge=merge)
